File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-5.java
class Solution {
    public int characterReplacement(String s, int k) {
        int res = 0;

        HashSet<Character> charSet = new HashSet<>();
        for (char c : s.toCharArray()) {
               charSet.add(c); 
        }

        for(char c : charSet) {
            int count= 0, lp = 0;
            for(int r=0;  r < s.length(); r++) {
                    if(s.charAt(r) == c) {
                        count++;
                    }

                    while((r -lp +1) - count > k ) {
                        if(s.charAt(lp) == c) {
                            count--;
                        }
                        lp++;
                    }

                    res = Math.max(res, (r-lp+1));
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-5.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:


        res = 0
        visited = set()

        l=r=0
        
        while(r < len(s)):
            
            if s[r] not in visited:
                visited.add(s[r])
                r+=1
            else:
                visited.remove(s[l])
                l+=1
            
            res = max(res, r-l)

        return res